### DISCUSSION TITLE
Black Box of BMS?

### DISCUSSION THREAD

**Discussion by Larsuu** [syssi/esphome-jbd-bms#66] [Ideas]
Hi

I just got my Esphome & JBD BMS (JBD-SP22S003A 40A 22S ) up and running through BLE. Going for a UART connection next. Still in a process of making my battery and now testing it in my shed. This JBD esphome really makes my battery even better. I have designed & built my very own battery heater controller and this jbdbms@esphome really makes it feel like deluxe battery with all the info and controls. 

It got me thinking that is there a way to save all the battery data when there is no network available? And then ofcourse sync it up to the home assistant to be properly backed up. 

Why? I was planning to do a data logging on my new battery, so:

- If one were to sell the battery it would have complete logs on how it was treated by the previous owner.

- If one would make li-ion batteries for sale, they would have complete data logging and thus easier detection of the problem.

- If Problem raises and causes havoc, one might have data to show what happened.

And now that i came to think of it, you could call it a black box. Then BLE would be better and have easier recover in case of thermal runaway. 

It gets me shivering a bit (mentally) everytime i see people selling random batteries with random mileage and unknown storing conditions. Could jbd-bms@esphome offer easy solution to this problem?

And yes, using a GPRS and tunneling it through VPN or some sort would do.. but this might be unreliable and not really scalable for general public.  

JBD BMS's app has some sort of logging.. is there a way to pull this?

**Comment by Larsuu** [comment]
Did some Googling and found that esphome does not support SD card. ( https://github.com/esphome/feature-requests/issues/513 )  So, it got me thinking that if there is a way of using flash and eeprom? Would it suffice?

Maybe eeprom memory could be used as saving details of the battery at the moment of emergency and then the previous records could be used along these in case of thermal runaway. So maybe Flash memory and eeprom could do if data logging is restricted to certain limits and only storing big and sudden changes in values? Depends on how much data there is flowing and how much would fit the flash. 

I doubt that all the output of each cell voltage can be stored in flash. Maybe just if it hits 10% above avarage? Isn't there esp32's with 32Mb of data? 

This sounds intresting.. i might get going it by myself later on, when my current project is ready. Just diggin for possible data on this topic.

**Reply by syssi** [reply]
Your idea is far behind the goal and supported features of ESPHome and HA. For example it's not possible to pass timestamped measurements to Home Assistant. All measurements are passed instantly and gets timestamped at the receiver (f.e. HA).

**Reply by Larsuu** [reply]
Thanks for your input. Yes, it's just a "prototype idea". It might need a raspberry or some sort of SBC. 

I have designed this Battery ECU for scooters and e-bikes and was thinking of adding more value to it with this idea.

But I might iterate the idea when I'm done with this current one. 

Thanks for the great work on this.
